### ISSUE TITLE
Bulk-Insert with EmptyString not working, also returns no error

### ISSUE THREAD

**Issue by olsn** [TimelordUK/node-sqlserver-v8#102]
Following issue:
When I do a bulk insert with an empty string, nothing will be written to the table, but response is that everything is fine - i'm using `node-mssql` with the `msnodesqlv8`-driver.

---

Code to reproduce:

```javascript
const config = require("../utils/config-loader").config;  // just db-credentials
const mssql = require("mssql/msnodesqlv8");

async function run() {
	await mssql.connect(config.mssqlConnectionString + "ETL_TEST");
	const sqlQuery = `
		CREATE TABLE [dbo].[TestEmptyString] (
				id NVARCHAR(12) NOT NULL,
				testfield NVARCHAR(12) NOT NULL,
				PRIMARY KEY (id)
		);`;
	await mssql.query(sqlQuery);
	
	const table = new mssql.Table(`[ETL_TEST].[dbo].[TestEmptyString]`);
	table.columns.add("id", mssql.NVarChar(12), {
		primary: true,
		nullable: false
	});
	table.columns.add("testfield", mssql.NVarChar(12), {
		primary: false,
		nullable: false
	});
	table.rows.add("1", "");
	const request = new mssql.Request();
	try {
		const res = await request.bulk(table);
		console.log(`Inserting data done - affected rows: ${res.rowsAffected}.`);
	} catch (e) {
		logger.error(e);
	}
	await mssql.close();
}

run();
```

When I set the column to nullable it works - but, it will then show `NULL` in the cell, which is not correct as an empty string is a valid value for a non-nullable column.

When trying to add the values via a plain SQL-Query everything works fine (`INSERT INTO ...`)

I was not able to find any issue in the JS-code after doing some extensive debugging.

**Comment by TimelordUK** [comment]
Is it the case it doesn’t work if we have a column with a mix of null and empty. If it is all null or all empty does it work for all rows? It may well be the cpp driver I will have to investigate further.

**Comment by olsn** [comment]
My example above inserts just a single row, so no there is no mix of any data. You should be able to use copy/paste the example and just exchange the `mssqlConnectionString` to reproduce the issue.

Test: `testfield NVARCHAR(12) NOT NULL`
- all rows emptyString -> does not work, no error
- all rows NULL -> does not work(but expected), no error
- mixed NULL and emptyString -> inserts the rows with emptyString correctly (so works, since NULL is not allowed)

Test: `testfield NVARCHAR(12) NULL`
- all rows emptyString -> adds as NULL instead of emptyString
- all rows NULL -> works
- mixed NULL and emptyString -> inserts the rows with emptyString and NULL correctly

In short: Mixed data works correctly, but if all rows have emptyString it does not behave correctly

**Comment by TimelordUK** [comment]
thanks for the investigation very helpful

- this was due to a fatal error in the c++ code causing a memory corruption.  I have fixed in 0.8.0

**Comment by olsn** [comment]
Awesome 👍  I'll test it either later today or tomorrow

**Comment by olsn** [comment]
Unfortunately the behavior in 0.8.0 is still the same for me - the code above to reproduce the issue still won't insert.

---

My current workaround is to add a null-row (all fields null) to my rows that I want to insert -> then everything works fine and the null-row will not be inserted, since my columns are "NOT NULL".

**Comment by TimelordUK** [comment]
Ok I’ll check once more I added a test at start of bulk test suite which essentially writes two rows both with empty string. Have I fixed a different problem?  The driver would crash on insert of bulk empty string in this case.  

I’ll run your test once more. Thanks.

**Comment by olsn** [comment]
That seems to be the exact problem, that you describe - maybe some publishing-issue?
I'll try to clean my cache + reinstall all node_modules.

**Comment by TimelordUK** [comment]
so i installed npm published version

i run the JS as below and console reports below

"C:\Program Files\nodejs\node.exe" C:\Users\Stephen\dev\js\ms8empty\index.js
Inserting data done - affected rows: 1.

Process finished with exit code 0

select * from TestEmptyString

and I see the below reported

id	testfield
1	




```javascript
let conf = require('./config/config')

// let's create the db configuration
const connectionConf = {
  user: '',
  password: '',
  database: conf.db.database,
  host: conf.db.host,
  connectionString: `Driver={${conf.db.optDriver}};Server=${conf.db.host};Database=${conf.db.database};Trusted_Connection=yes;`,
  options: {
    driver: conf.db.optDriver,
    trustedConnection: true,
    instanceName: conf.db.optInstanceName
  },

  pool: {
    max: 5,
    min: 0,
    idle: 10000
  }
}
const mssql = require("mssql/msnodesqlv8");

async function run() {
  await mssql.connect(connectionConf.connectionString);
  const sqlQuery = `
		CREATE TABLE [dbo].[TestEmptyString] (
				id NVARCHAR(12) NOT NULL,
				testfield NVARCHAR(12) NOT NULL,
				PRIMARY KEY (id)
		);`;
  // await mssql.query(sqlQuery);

  const table = new mssql.Table(`[scratch].[dbo].[TestEmptyString]`);
  table.columns.add("id", mssql.NVarChar(12), {
    primary: true,
    nullable: false
  });
  table.columns.add("testfield", mssql.NVarChar(12), {
    primary: false,
    nullable: false
  });
  table.rows.add("1", "");
  const request = new mssql.Request();
  try {
    const res = await request.bulk(table);
    console.log(`Inserting data done - affected rows: ${res.rowsAffected}.`);
  } catch (e) {
    logger.error(e);
  }
  await mssql.close();
}

run();

```

**Comment by TimelordUK** [comment]
out of interest which version of node are you on - note i only publish binaries for versions 10,11 and 12 - could you be picking up an old version from somewhere else.

**Comment by olsn** [comment]
I'm using Node v10.16.0
I cleaned the npm cache, removed the node_modules, remove the package-lock.json, did an npm install.
I have verified, that 0.8.0 is installed directly in the node_modules.
It even says `Inserting data done - affected rows: 1.`
But the table in the DB is empty.

Is there some flag that I can use to enable verbose logging or something like this?

**Comment by TimelordUK** [comment]
Can you select in your test app is it possible if you are using a client to look at the table that the library is writing to a different schema with same name.  In other words after the write where it says one row effected have you got a select.  

I dont really have that much logging. I’m wondering if the sql created in table manager is actually targeting correct schema or not.

**Comment by olsn** [comment]
It should be the correct schema, because when I add a non-emtpy-string it works and actually adds it to that table.
I'm using the default dbo-schema.

> Can you select in your test app is it possible if you are using a client to look at the table that the library is writing to a different schema with same name. In other words after the write where it says one row effected have you got a select. I dont really have that much logging. I’m wondering if the sql created in table manager is actually targeting correct schema or not.
> […](#)
> Sent from my iPhone
> On 11 Jul 2019, at 10:12, Olaf Horstmann ***@***.***> wrote: I'm using Node v10.16.0 I cleaned the npm cache, removed the node_modules, remove the package-lock.json, did an npm install. I have verified, that 0.8.0 is installed directly in the node_modules. It even says Inserting data done - affected rows: 1. But the table in the DB is empty. Is there some flag that I can use to enable verbose logging or something like this? — You are receiving this because you commented. Reply to this email directly, view it on GitHub, or mute the thread.

**Comment by TimelordUK** [comment]
I am closing. I believe this is fixed. Please provide working sample on latest version that reproduces issue as we now have unit tests for this and null insert.
